modyn/trainer_server/internal/trainer/remote_downsamplers/remote_craig_downsampling.py

```diff
@@ -0,0 +1,196 @@
+    """
+    Strategy introduced in:
+    Data-efficient Training of Machine Learning Models
+    Implementation adapted from:
+    DEEPCORE https://raw.githubusercontent.com/PatrickZH/DeepCore/main/deepcore/methods/craig.py
+    This strategy selects points via submodular maximization of a per-class FacilityLocation function. The score is
+    proportional to the Euclidean distance between the two samples' gradients.
+    """
+
+    def __init__(
+        self,
+        pipeline_id: int,
+        trigger_id: int,
+        batch_size: int,
+        params_from_selector: dict,
+        per_sample_loss: Any,
+        device: str,
+    ) -> None:
+        super().__init__(pipeline_id, trigger_id, batch_size, params_from_selector, device)
+
+        self.criterion = per_sample_loss
+
+        self.selection_batch = params_from_selector["selection_batch"]
+        self.greedy = params_from_selector["greedy"]
+        if self.greedy not in OPTIMIZER_CHOICES:
+            raise ValueError(
+                f"The required Greedy optimizer is not available. Pick one of the following: {OPTIMIZER_CHOICES}"
+            )
+
+        # This class uses the embedding recorder
+        self.requires_coreset_supporting_module = True
+
+        # Samples are supplied label by label (this class is instance of AbstractPerLabelRemoteDownsamplingStrategy).
+        # The following list keeps the gradients of the current label. When all the samples belonging to the current
+        # label have been seen, the scores are computed and the list is emptied
+        self.current_class_gradients: list[torch.Tensor] = []
+        # distance_matrix[i,j] = 0 if label[i]!=label[j] else is proportional to the euclidean distance between the
+        # two samples in the gradient space
+        self.distance_matrix = np.zeros([0, 0])
+
+        # if true, the downsampling is balanced across classes ex class sizes = [10, 50, 30] and 50% downsampling
+        # yields the following downsampled class sizes [5, 25, 15] while without balance something like [0, 45, 0] can
+        # happen
+        self.balance = params_from_selector["balance"]
+        if self.balance:
+            self.already_selected_samples: list[int] = []
+            self.already_selected_weights = torch.tensor([]).float()
+
+    def inform_samples(
+        self,
+        sample_ids: list[int],
+        forward_output: torch.Tensor,
+        target: torch.Tensor,
+        embedding: Optional[torch.Tensor] = None,
+    ) -> None:
+        assert embedding is not None
+
+        # Slightly different implementation for BTS and STB since in STB points are supplied class by class while in
+        # BTS are not. STB will always use the first branch, STB will typically (might use the first if all the points
+        # belong to the same class) use the second one
+        different_targets_in_this_batch = target.unique()
+        if len(different_targets_in_this_batch) == 1:
```

Comment:
Is this your proxy for STB vs BTS? Because what if we only have samples of one class in BTS?

Reply:
The problem with CRAIG is that the algorithm considers class when selecting points. This is not a problem in STB since there is a dedicated dataloader to receive the samples as desired.
In BTS it is different because potentially in the same batch, there could be samples from different classes. If this happens, the algorithm is used in a different way (receiving samples class by class, emulating what happens in STB).

It is not a proxy for STB vs. BTS because there is no problem if there are samples from a single class in BTS. It is just to handle the special case of multiple classes in the same batch.

Reply:
Ahh, could you add this as a comment here as well? Keeping this knowledge for the future. Thank you!